cmd/util/ledger/migrations/cadence_values_migration.go

```diff
@@ -0,0 +1,243 @@
+	}
+}
+
+func (m *CadenceValueMigrator) Close() error {
+	// Close the report writer so it flushes to file.
+	m.reporter.Close()
+	return nil
+}
+
+func (m *CadenceValueMigrator) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	_ int,
+) error {
+	m.log = log.With().Str("migration", "cadence-value-migration").Logger()
+	return nil
+}
+
+func (m *CadenceValueMigrator) MigrateAccount(
+	_ context.Context,
+	address common.Address,
+	oldPayloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+
+	// Create all the runtime components we need for the migration
+	migrationRuntime, err := newMigratorRuntime(address, oldPayloads)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create migrator runtime: %w", err)
+	}
+
+	migration := migrations.NewStorageMigration(
+		migrationRuntime.Interpreter,
+		migrationRuntime.Storage,
+	)
+
+	capabilityIDs := map[interpreter.AddressPath]interpreter.UInt64Value{}
+
+	reporter := newValueMigrationReporter(m.reporter)
+
+	m.log.Info().Msg("Migrating cadence values")
+
+	migration.Migrate(
+		&migrations.AddressSliceIterator{
+			Addresses: []common.Address{
+				address,
+			},
+		},
+		migration.NewValueMigrationsPathMigrator(
+			reporter,
+			&capcons.CapabilityValueMigration{
+				CapabilityIDs: capabilityIDs,
+				Reporter:      reporter,
+			},
+			&capcons.LinkValueMigration{
+				CapabilityIDs:      capabilityIDs,
+				AccountIDGenerator: &AccountIDGenerator{},
+				Reporter:           reporter,
+			},
+			string_normalization.NewStringNormalizingMigration(),
+			account_type.NewAccountTypeMigration(),
+			type_value.NewTypeValueMigration(),
+		),
+	)
+
+	m.log.Info().Msg("Committing changes")
+	migration.Commit()
+
+	// finalize the transaction
+	result, err := migrationRuntime.TransactionState.FinalizeMainTransaction()
+	if err != nil {
+		return nil, fmt.Errorf("failed to finalize main transaction: %w", err)
+	}
+
+	// Merge the changes to the original payloads.
+	return m.mergeRegisterChanges(migrationRuntime, result.WriteSet)
+}
+
+func (m *CadenceValueMigrator) mergeRegisterChanges(
+	mr *migratorRuntime,
+	changes map[flow.RegisterID]flow.RegisterValue,
+) ([]*ledger.Payload, error) {
+
+	originalPayloads := mr.Snapshot.Payloads
+	newPayloads := make([]*ledger.Payload, 0, len(originalPayloads))
+
+	for id, value := range originalPayloads {
+		if len(value.Value()) == 0 {
+			// This is strange, but we don't want to add empty values. Log it.
+			m.log.Warn().Msgf("empty value for key %s", id)
+			continue
+		}
+
+		// If the payload had changed, then use the updated payload.
+		if updatedPayload, contains := changes[id]; contains {
+			key := convert.RegisterIDToLedgerKey(id)
+			newPayloads = append(newPayloads, ledger.NewPayload(key, updatedPayload))
+		} else {
+			newPayloads = append(newPayloads, value)
+		}
+	}
+
+	return newPayloads, nil
+}
+
+// AccountIDGenerator for the link-value migration
+type AccountIDGenerator struct {
+	ids map[common.Address]uint64
```

Comment:
after the migration is complete these would need to be written to the account state, otherwise the accounts will still be at ID 0 once the network starts up.

I think you need to create an `NewAccountLocalIDGenerator` using the `environment.NewAccounts` that are actually created during `newMigratorRuntime` but are not yet exposed.

Reply:
Replaced with `NewAccountLocalIDGenerator` [361727d](https://github.com/onflow/flow-go/pull/5192/commits/361727d38b804354d6a13ad16cbc80d3c19f6c8d).